chromadb/api/configuration.py

```diff
@@ -0,0 +1,228 @@
+                )
+
+    def __repr__(self) -> str:
+        return f"Configuration({self.parameter_map.values()})"
+
+    def __eq__(self, __value: object) -> bool:
+        if not isinstance(__value, Configuration):
+            return NotImplemented
+        return self.parameter_map == __value.parameter_map
+
+    def get_parameters(self) -> List[ConfigurationParameter]:
+        """Returns the parameters of the configuration."""
+        return list(self.parameter_map.values())
+
+    def get_parameter(self, name: str) -> Optional[ConfigurationParameter]:
+        """Returns the parameter with the given name or None if it does not exist."""
+        return self.parameter_map.get(name, None)
+
+    def set_parameter(self, name: str, value: Union[str, int, float, bool]) -> None:
+        """Sets the parameter with the given name to the given value."""
+        if name not in self.definitions:
+            raise ValueError(f"Invalid parameter name: {name}")
+        definition = self.definitions[name]
+        parameter = self.parameter_map[name]
+        if definition.is_static:
+            raise StaticParameterError(f"Cannot set static parameter: {name}")
+        if not definition.validator(value):
+            raise ValueError(f"Invalid value for parameter {name}: {value}")
+        parameter.value = value
+
+    @override
+    def to_json_str(self) -> str:
+        """Returns the JSON representation of the configuration."""
+        return json.dumps(self.to_json())
+
+    @classmethod
+    @override
+    def from_json_str(cls, json_str: str) -> Self:
+        """Returns a configuration from the given JSON string."""
+        try:
+            config_json = json.loads(json_str)
+        except json.JSONDecodeError:
+            raise ValueError(
+                f"Unable to decode configuration from JSON string: {json_str}"
+            )
+        return cls.from_json(config_json)
+
+    @override
+    def to_json(self) -> Dict[str, Any]:
+        """Returns the JSON compatible dictionary representation of the configuration."""
+        return {name: parameter.value for name, parameter in self.parameter_map.items()}
+
+    @classmethod
+    @override
+    def from_json(cls, json_map: Dict[str, Any]) -> Self:
+        """Returns a configuration from the given JSON string."""
+        parameters = []
+        for name, value in json_map.items():
+            parameters.append(ConfigurationParameter(name=name, value=value))
+        return cls(parameters=parameters)
+
+
+class CollectionConfiguration(Configuration):
+    """The configuration for a collection."""
+
+    definitions = {
+        "space": ConfigurationDefinition(
+            name="space",
+            validator=lambda value: isinstance(value, str)
+            and value in ["l2", "ip", "cosine"],
+            is_static=True,
+            default_value="l2",
+        ),
+        "ef_construction": ConfigurationDefinition(
```

Comment:
I wonder if we want to prefix these in any way? Imagine a world with more than just hnsw indices, how could we make that clear?

Reply:
We should, this was a miss on my part. `hnsw:` seems correct / safe? I could do that in this PR, or else in the follow-up where we actually consume this. Doing it now might save a migration.

Reply:
Will address this in a PR above this one.